src/RecordManager/Base/Harvest/GeniePlus.php

```diff
@@ -0,0 +1,682 @@
+        }
+    }
+
+    /**
+     * Create an OAI style ID
+     *
+     * @param string $sourceId Source ID
+     * @param string $id       Record ID
+     *
+     * @return string OAI ID
+     */
+    protected function createOaiId($sourceId, $id)
+    {
+        return "genieplus:$sourceId:$id";
+    }
+
+    /**
+     * Extract display values from an API response field.
+     *
+     * @param array $field Array of values from API
+     *
+     * @return array
+     */
+    protected function extractDisplayValues($field): array
+    {
+        $callback = function ($value) {
+            return $value['display'];
+        };
+        return array_map($callback, $field);
+    }
+
+    /**
+     * Extract holdings data from an API response. Return an array of arrays
+     * representing 852 fields (indexed by subfield code).
+     *
+     * @param array $record Record from API response
+     *
+     * @return array
+     */
+    protected function getHoldings($record): array
+    {
+        $locations = $this
+            ->extractDisplayValues($record[$this->locationField] ?? []);
+        $sublocations = $this
+            ->extractDisplayValues($record[$this->sublocationField] ?? []);
+        $callNos = $this
+            ->extractDisplayValues($record[$this->callnumberField] ?? []);
+        $barcodes = $this->extractDisplayValues($record[$this->barcodeField] ?? []);
+        $total = max(
+            [
+                count($locations),
+                count($sublocations),
+                count($callNos),
+                count($barcodes),
+            ]
+        );
+        $result = [];
+        for ($i = 0; $i < $total; $i++) {
+            $result[] = [
+                'a' => $locations[$i] ?? '',
+                'b' => $sublocations[$i] ?? '',
+                'h' => $callNos[$i] ?? '',
+                'p' => $barcodes[$i] ?? '',
+            ];
+        }
+        return $result;
+    }
+
+    /**
+     * Extract/format MARC record found in API response
+     *
+     * @param array $record GeniePlus record from API response
+     *
+     * @return array
+     */
+    protected function processMarcRecord($record)
+    {
+        if (!isset($record[$this->marcField][0]['display'])) {
+            throw new \Exception("Missing MARC field: {$this->marcField}");
+        }
+        // Extract MARC and ID from API response
+        $marc = $record[$this->marcField][0]['display'];
+        $id = $record[$this->idField][0]['display'];
+
+        // Convert to XML (and strip any illegal characters that slipped through)
+        $replacementCount = 0;
+        $rawXml = preg_replace(
+            '/[^\x{0009}\x{000a}\x{000d}\x{0020}-\x{D7FF}\x{E000}-\x{FFFD}]+/u',
+            '',
+            $this->lineBasedFormatter->convertLineBasedMarcToXml($marc),
+            -1,
+            $replacementCount
+        );
+        if ($replacementCount > 0) {
+            $this->infoMsg("Replaced $replacementCount bad chars in record $id");
```

Comment:
infoMsg may be fine, but consider warningMsg/debugMsg as well depending on how frequent occurrence this is and how much bearing it has on the resulting records.

Reply:
Yes, I think warning is better -- changed (along with merging in #98).